File: packages/sdk-go/locator.go
package stagehand

import (
	"context"
	"encoding/base64"
	"fmt"
	"os"
	"path/filepath"
)

// FileInput describes either a local file path or an in-memory file payload.
// Use FilePath or FileData to construct one.
type FileInput struct {
	Path         string
	Name         string
	MIMEType     string
	Buffer       []byte
	LastModified *int
}

// FilePath creates an upload from a path on the SDK caller's filesystem.
func FilePath(path string) FileInput {
	return FileInput{Path: path}
}

// FileData creates an in-memory upload payload.
func FileData(name string, mimeType string, buffer []byte) FileInput {
	return FileInput{Name: name, MIMEType: mimeType, Buffer: append([]byte(nil), buffer...)}
}

// PageLocator is the client wrapper named Locator in the TypeScript and Python
// SDKs. The Go protocol already exports a different generated Locator value.
type PageLocator struct {
	rpc        protocolClient
	descriptor LocatorDescriptor
}

// Descriptor returns the generated wire descriptor.
func (l *PageLocator) Descriptor() LocatorDescriptor {
	return l.descriptor
}

// Click clicks the matching element.
func (l *PageLocator) Click(ctx context.Context, options *LocatorClickOptions) error {
	params := LocatorClickParams{
		PageID: l.descriptor.PageID, Selector: l.descriptor.Selector, Nth: l.descriptor.Nth,
		Options: options,
	}
	var result LocatorClickResult
	return l.rpc.call(ctx, "locator.click", params, &result)
}

// Hover hovers the matching element.
func (l *PageLocator) Hover(ctx context.Context) error {
	params := l.descriptor
	var result LocatorHoverResult
	return l.rpc.call(ctx, "locator.hover", params, &result)
}

// Fill replaces the matching input's value.
func (l *PageLocator) Fill(ctx context.Context, value string) error {
	params := LocatorFillParams{
		PageID: l.descriptor.PageID, Selector: l.descriptor.Selector, Nth: l.descriptor.Nth,
		Value: value,
	}
	var result LocatorFillResult
	return l.rpc.call(ctx, "locator.fill", params, &result)
}

// Count returns the number of matching elements.
func (l *PageLocator) Count(ctx context.Context) (int, error) {
	params := l.descriptor
	var result LocatorCountResult
	if err := l.rpc.call(ctx, "locator.count", params, &result); err != nil {
		return 0, err
	}
	return int(result), nil
}

// IsChecked reports whether the matching control is checked.
func (l *PageLocator) IsChecked(ctx context.Context) (bool, error) {
	params := l.descriptor
	var result LocatorIsCheckedResult
	if err := l.rpc.call(ctx, "locator.is_checked", params, &result); err != nil {
		return false, err
	}
	return bool(result), nil
}

// InputValue returns the matching input's value.
func (l *PageLocator) InputValue(ctx context.Context) (string, error) {
	params := l.descriptor
	var result LocatorInputValueResult
	if err := l.rpc.call(ctx, "locator.input_value", params, &result); err != nil {
		return "", err
	}
	return string(result), nil
}

// IsVisible reports whether the matching element is visible.
func (l *PageLocator) IsVisible(ctx context.Context) (bool, error) {
	params := l.descriptor
	var result LocatorIsVisibleResult
	if err := l.rpc.call(ctx, "locator.is_visible", params, &result); err != nil {
		return false, err
	}
	return bool(result), nil
}

// InnerText returns the matching element's rendered text.
func (l *PageLocator) InnerText(ctx context.Context) (string, error) {
	params := l.descriptor
	var result LocatorInnerTextResult
	if err := l.rpc.call(ctx, "locator.inner_text", params, &result); err != nil {
		return "", err
	}
	return string(result), nil
}

// InnerHTML returns the matching element's HTML.
func (l *PageLocator) InnerHTML(ctx context.Context) (string, error) {
	params := l.descriptor
	var result LocatorInnerHTMLResult
	if err := l.rpc.call(ctx, "locator.inner_html", params, &result); err != nil {
		return "", err
	}
	return string(result), nil
}

// TextContent returns the matching element's text content.
func (l *PageLocator) TextContent(ctx context.Context) (string, error) {
	params := l.descriptor
	var result LocatorTextContentResult
	if err := l.rpc.call(ctx, "locator.text_content", params, &result); err != nil {
		return "", err
	}
	return string(result), nil
}

// ScrollTo scrolls the matching element to a generated percentage value.
func (l *PageLocator) ScrollTo(ctx context.Context, percent ScrollPercent) error {
	params := LocatorScrollToParams{
		PageID: l.descriptor.PageID, Selector: l.descriptor.Selector, Nth: l.descriptor.Nth,
		Percent: percent,
	}
	var result LocatorScrollToResult
	return l.rpc.call(ctx, "locator.scroll_to", params, &result)
}

// Centroid returns the matching element's center coordinates.
func (l *PageLocator) Centroid(ctx context.Context) (LocatorCentroidResult, error) {
	params := l.descriptor
	var result LocatorCentroidResult
	err := l.rpc.call(ctx, "locator.centroid", params, &result)
	return result, err
}

// Highlight highlights the matching element.
func (l *PageLocator) Highlight(ctx context.Context, options *LocatorHighlightOptions) error {
	params := LocatorHighlightParams{
		PageID: l.descriptor.PageID, Selector: l.descriptor.Selector, Nth: l.descriptor.Nth,
		Options: options,
	}
	var result LocatorHighlightResult
	return l.rpc.call(ctx, "locator.highlight", params, &result)
}

// SendClickEvent sends a click event to the matching element.
func (l *PageLocator) SendClickEvent(
	ctx context.Context,
	options *LocatorSendClickEventOptions,
) error {
	params := LocatorSendClickEventParams{
		PageID: l.descriptor.PageID, Selector: l.descriptor.Selector, Nth: l.descriptor.Nth,
		Options: options,
	}
	var result LocatorSendClickEventResult
	return l.rpc.call(ctx, "locator.send_click_event", params, &result)
}

// Type enters text into the matching element.
func (l *PageLocator) Type(ctx context.Context, text string, options *LocatorTypeOptions) error {
	params := LocatorTypeParams{
		PageID: l.descriptor.PageID, Selector: l.descriptor.Selector, Nth: l.descriptor.Nth,
		Text: text, Options: options,
	}
	var result LocatorTypeResult
	return l.rpc.call(ctx, "locator.type", params, &result)
}

// SelectOption selects values in the matching element.
func (l *PageLocator) SelectOption(ctx context.Context, values StringList) ([]string, error) {
	params := LocatorSelectOptionParams{
		PageID: l.descriptor.PageID, Selector: l.descriptor.Selector, Nth: l.descriptor.Nth,
		Values: values,
	}
	var result LocatorSelectOptionResult
	if err := l.rpc.call(ctx, "locator.select_option", params, &result); err != nil {
		return nil, err
	}
	return []string(result), nil
}

// SetInputFiles sets files on the matching <input type="file"> element. Calling
// it without files clears the current selection.
func (l *PageLocator) SetInputFiles(ctx context.Context, files ...FileInput) error {
	payloads := make([]InputFilePayload, 0, len(files))
	for _, file := range files {
		payload, err := normalizeFileInput(file)
		if err != nil {
			return err
		}
		payloads = append(payloads, payload)
	}

	params := LocatorSetInputFilesParams{
		PageID: l.descriptor.PageID, Selector: l.descriptor.Selector, Nth: l.descriptor.Nth,
		Files: payloads,
	}
	var result LocatorSetInputFilesResult
	return l.rpc.call(ctx, "locator.set_input_files", params, &result)
}

func normalizeFileInput(file FileInput) (InputFilePayload, error) {
	if file.Path != "" {
		if file.Name != "" || file.MIMEType != "" || file.Buffer != nil || file.LastModified != nil {
			return InputFilePayload{}, fmt.Errorf("set input files: path and payload fields cannot be combined")
		}
		absolutePath, err := filepath.Abs(file.Path)
		if err != nil {
			return InputFilePayload{}, fmt.Errorf("set input files: resolve %q: %w", file.Path, err)
		}
		info, err := os.Stat(absolutePath)
		if err != nil {
			return InputFilePayload{}, fmt.Errorf("set input files: stat %q: %w", absolutePath, err)
		}
		if !info.Mode().IsRegular() {
			return InputFilePayload{}, fmt.Errorf("set input files: expected a file at %q", absolutePath)
		}
		data, err := os.ReadFile(absolutePath)
		if err != nil {
			return InputFilePayload{}, fmt.Errorf("set input files: read %q: %w", absolutePath, err)
		}
		lastModified := int(info.ModTime().UnixMilli())
		return InputFilePayload{
			Name: filepath.Base(absolutePath), Data: base64.StdEncoding.EncodeToString(data),
			LastModified: &lastModified,
		}, nil
	}

	if file.Name == "" {
		return InputFilePayload{}, fmt.Errorf("set input files: payload name cannot be empty")
	}
	payload := InputFilePayload{
		Name: file.Name, Data: base64.StdEncoding.EncodeToString(file.Buffer),
		LastModified: file.LastModified,
	}
	if file.MIMEType != "" {
		payload.MIMEType = &file.MIMEType
	}
	return payload, nil
}

// First returns a locator restricted to the first match.
func (l *PageLocator) First() *PageLocator {
	index := 0
	descriptor := l.descriptor
	descriptor.Nth = &index
	return &PageLocator{rpc: l.rpc, descriptor: descriptor}
}

// Nth returns a locator restricted to one zero-based match. It returns an
// error when index is negative.
func (l *PageLocator) Nth(index int) (*PageLocator, error) {
	if index < 0 {
		return nil, fmt.Errorf("stagehand locator index must be non-negative: %d", index)
	}
	descriptor := l.descriptor
	descriptor.Nth = &index
	return &PageLocator{rpc: l.rpc, descriptor: descriptor}, nil
}

File: packages/sdk-python/src/stagehand/locator.py
from __future__ import annotations

from collections.abc import Sequence
from typing import Literal, Self, cast

from ._generated.input_types import RgbaColor
from ._generated.models import (
    LocatorCentroidResult,
    LocatorClickOptions,
    LocatorClickParams,
    LocatorClickResult,
    LocatorCountResult,
    LocatorDescriptor,
    LocatorFillParams,
    LocatorFillResult,
    LocatorHighlightOptions,
    LocatorHighlightParams,
    LocatorHighlightResult,
    LocatorHoverResult,
    LocatorInnerHtmlResult,
    LocatorInnerTextResult,
    LocatorInputValueResult,
    LocatorIsCheckedResult,
    LocatorIsVisibleResult,
    LocatorOperationParams,
    LocatorScrollToParams,
    LocatorScrollToResult,
    LocatorSelectOptionParams,
    LocatorSelectOptionResult,
    LocatorSendClickEventOptions,
    LocatorSendClickEventParams,
    LocatorSendClickEventResult,
    LocatorSetInputFilesParams,
    LocatorSetInputFilesResult,
    LocatorTextContentResult,
    LocatorTypeOptions,
    LocatorTypeParams,
    LocatorTypeResult,
    MouseButton,
)
from .file_upload import FileInput, normalize_file_input
from .rpc_client import RPCClient


class Locator:
    def __init__(
        self,
        rpc_client: RPCClient,
        *,
        page_id: str,
        selector: str,
        nth: int | None = None,
    ) -> None:
        self._rpc_client = rpc_client
        self._descriptor = LocatorDescriptor(page_id=page_id, selector=selector)
        if nth is not None:
            self._descriptor.nth = nth

    @property
    def page_id(self) -> str:
        return self._descriptor.page_id

    @property
    def selector(self) -> str:
        return self._descriptor.selector

    @property
    def nth_index(self) -> int | None:
        return self._descriptor.nth

    @property
    def descriptor(self) -> LocatorDescriptor:
        return self._descriptor

    async def click(
        self,
        *,
        button: MouseButton | Literal["left", "right", "middle"] | None = None,
        click_count: int | None = None,
        timeout: int | None = None,
    ) -> None:
        values = self._descriptor.model_dump(exclude_unset=True)
        options = LocatorClickOptions.model_validate({
            name: value
            for name, value in (
                ("button", button),
                ("click_count", click_count),
                ("timeout", timeout),
            )
            if value is not None
        })
        if options.model_fields_set:
            values["options"] = options
        await self._rpc_client.send(
            "locator.click",
            LocatorClickParams.model_validate(values),
            LocatorClickResult,
        )

    async def hover(self, *, timeout: int | None = None) -> None:
        await self._rpc_client.send(
            "locator.hover",
            LocatorOperationParams.model_validate(self._operation_values(timeout)),
            LocatorHoverResult,
        )

    async def fill(self, value: str, *, timeout: int | None = None) -> None:
        await self._rpc_client.send(
            "locator.fill",
            LocatorFillParams.model_validate({
                **self._operation_values(timeout),
                "value": value,
            }),
            LocatorFillResult,
        )

    async def count(self, *, timeout: int | None = None) -> int:
        return await self._rpc_client.send(
            "locator.count",
            LocatorOperationParams.model_validate(self._operation_values(timeout)),
            LocatorCountResult,
        )

    async def is_checked(self, *, timeout: int | None = None) -> bool:
        return await self._rpc_client.send(
            "locator.is_checked",
            LocatorOperationParams.model_validate(self._operation_values(timeout)),
            LocatorIsCheckedResult,
        )

    async def input_value(self, *, timeout: int | None = None) -> str:
        return await self._rpc_client.send(
            "locator.input_value",
            LocatorOperationParams.model_validate(self._operation_values(timeout)),
            LocatorInputValueResult,
        )

    async def is_visible(self, *, timeout: int | None = None) -> bool:
        return await self._rpc_client.send(
            "locator.is_visible",
            LocatorOperationParams.model_validate(self._operation_values(timeout)),
            LocatorIsVisibleResult,
        )

    async def inner_text(self, *, timeout: int | None = None) -> str:
        return await self._rpc_client.send(
            "locator.inner_text",
            LocatorOperationParams.model_validate(self._operation_values(timeout)),
            LocatorInnerTextResult,
        )

    async def inner_html(self, *, timeout: int | None = None) -> str:
        return await self._rpc_client.send(
            "locator.inner_html",
            LocatorOperationParams.model_validate(self._operation_values(timeout)),
            LocatorInnerHtmlResult,
        )

    async def text_content(self, *, timeout: int | None = None) -> str:
        return await self._rpc_client.send(
            "locator.text_content",
            LocatorOperationParams.model_validate(self._operation_values(timeout)),
            LocatorTextContentResult,
        )

    async def scroll_to(self, percent: float | str, *, timeout: int | None = None) -> None:
        await self._rpc_client.send(
            "locator.scroll_to",
            LocatorScrollToParams.model_validate({
                **self._operation_values(timeout),
                "percent": percent,
            }),
            LocatorScrollToResult,
        )

    async def centroid(self, *, timeout: int | None = None) -> LocatorCentroidResult:
        return await self._rpc_client.send(
            "locator.centroid",
            LocatorOperationParams.model_validate(self._operation_values(timeout)),
            LocatorCentroidResult,
        )

    async def highlight(
        self,
        *,
        duration_ms: int | None = None,
        border_color: RgbaColor | None = None,
        content_color: RgbaColor | None = None,
        timeout: int | None = None,
    ) -> None:
        values = self._descriptor.model_dump(exclude_unset=True)
        options = LocatorHighlightOptions.model_validate({
            name: value
            for name, value in (
                ("duration_ms", duration_ms),
                ("border_color", border_color),
                ("content_color", content_color),
                ("timeout", timeout),
            )
            if value is not None
        })
        if options.model_fields_set:
            values["options"] = options
        await self._rpc_client.send(
            "locator.highlight",
            LocatorHighlightParams.model_validate(values),
            LocatorHighlightResult,
        )

    async def send_click_event(
        self,
        *,
        bubbles: bool | None = None,
        cancelable: bool | None = None,
        composed: bool | None = None,
        detail: float | None = None,
        timeout: int | None = None,
    ) -> None:
        values = self._descriptor.model_dump(exclude_unset=True)
        options = LocatorSendClickEventOptions.model_validate({
            name: value
            for name, value in (
                ("bubbles", bubbles),
                ("cancelable", cancelable),
                ("composed", composed),
                ("detail", detail),
                ("timeout", timeout),
            )
            if value is not None
        })
        if options.model_fields_set:
            values["options"] = options
        await self._rpc_client.send(
            "locator.send_click_event",
            LocatorSendClickEventParams.model_validate(values),
            LocatorSendClickEventResult,
        )

    async def type(
        self, text: str, *, delay: float | None = None, timeout: int | None = None
    ) -> None:
        values = {**self._operation_values(timeout), "text": text}
        options = LocatorTypeOptions.model_validate({
            name: value
            for name, value in (("delay", delay), ("timeout", timeout))
            if value is not None
        })
        if options.model_fields_set:
            values["options"] = options
        await self._rpc_client.send(
            "locator.type",
            LocatorTypeParams.model_validate(values),
            LocatorTypeResult,
        )

    async def select_option(
        self, values: str | Sequence[str], *, timeout: int | None = None
    ) -> list[str]:
        return await self._rpc_client.send(
            "locator.select_option",
            LocatorSelectOptionParams.model_validate({
                **self._operation_values(timeout),
                "values": list(values) if not isinstance(values, str) else values,
            }),
            LocatorSelectOptionResult,
        )

    async def set_input_files(
        self,
        files: FileInput | Sequence[FileInput],
        *,
        timeout: int | None = None,
    ) -> None:
        await self._rpc_client.send(
            "locator.set_input_files",
            LocatorSetInputFilesParams.model_validate({
                **self._operation_values(timeout),
                "files": normalize_file_input(files),
            }),
            LocatorSetInputFilesResult,
        )

    def _operation_values(self, timeout: int | None) -> dict[str, object]:
        values = self._descriptor.model_dump(exclude_unset=True)
        if timeout is not None:
            values["options"] = {"timeout": timeout}
        return cast(dict[str, object], values)

    def first(self) -> Self:
        return self.nth(0)

    def nth(self, index: int) -> Self:
        return type(self)(
            self._rpc_client,
            page_id=self.page_id,
            selector=self.selector,
            nth=index,
        )
